identity-service/src/routes/trackListens.js

```diff
@@ -203,15 +207,80 @@ module.exports = function (app) {
     let currentHour = await getListenHour()
     let trackListenRecord = await models.TrackListenCount.findOrCreate(
       {
-        where: { hour: currentHour, trackId: req.params.id }
+        where: { hour: currentHour, trackId }
       })
     if (trackListenRecord && trackListenRecord[1]) {
       logger.info(`New track listen record inserted ${trackListenRecord}`)
     }
     await models.TrackListenCount.increment('listens', { where: { hour: currentHour, trackId: req.params.id } })
+
+    // Find / Create the record of the user listening to the track
+    const [userTrackListenRecord, created] = await models.UserTrackListen
+      .findOrCreate({ where: { userId: req.body.userId, trackId } })
+
+    // If the recrod was not created, updated the timestamp
+    if (!created) {
+      userTrackListenRecord.set('updatedAt', new Date())
+      await userTrackListenRecord.save()
+    }
+
+    // Count the total number of tracks listens to by the user
+    const userTrackListenCount = await models.UserTrackListen.count({
+      where: { userId: req.body.userId }
+    })
+
+    // Remove the oldest tracks if the count is greater than the limit
+    if (userTrackListenCount > USER_TRACK_LISTEN_COUNT_LIMIT) {
```

Comment:
@jowlee i feel like a lot of these are a lot of SQL queries for every time you want to log a listen. for every listen we're doing 4 (and sometimes 5 sql queries) on the pool, not even in a single tx. i wonder if we can cut the number down or use some kind of stored proc. I'd almost rather not prune the track counts > 1000 here and do it on some other job